design/bgp-bfd.md

```diff
@@ -0,0 +1,157 @@
+# BFD enablement in BGP
+
+## Summary
+
+The purpose of this enhancement is to use enable [BFD](https://en.wikipedia.org/wiki/Bidirectional_Forwarding_Detection) to be used in conjunction with BGP.
+
+BFD will be available only in the upcoming [FRR based implementation](0001-frr.md).
+
+## Motivation
+
+BFD provides a quicker path failure detection than BGP, and using them together will allow users to provide a better
+service.
+
+BFD for BGP is supported by FRR [out of the box](http://docs.frrouting.org/en/latest/bfd.html#bgp-bfd-configuration).
+
+### Goals
+
+* Provide a way to establish BFD on BGP session.
+* Provide a way to configure the parameters of the BFD session associated with a BGP session.
+
+### Non-Goals
+
+* Implement BFD on the legacy BGP implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to declare a BGP session to be backed up by BFD.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to set the BFD parameters related to a BGP session.
+
+## Design Details
+
+### BFD Design proposal
+
+The idea is to leverage FRR to enable BFD on a BGP session.
+
+When declaring a BFD peer, all it takes to enable BFD is to add a `bfd` property on the neighbour:
+
+```yaml
+    neighbor <A.B.C.D|X:X::X:X|WORD> bfd profile BFDPROF
+```
+
+The proposal here is to define a bfd profile section in the config structure that looks like:
+
+```yaml
+    bfd-profiles:
```

Comment:
FRR has a few more profile configuration options available (e.g. `minimum-ttl`, `echo-mode`, `echo receive-interval`, `echo transmit-interval`). Should those also be supported in the BFD config structure for MetalLB?

The `minimum-ttl` may be potentially required for multi-hop sessions (e.g. BGP peering with DC gateway or border leaf, rather than with ToR switch) because the default value in FRR only allows one hop between the system and its peers).

Reply:
Uhm, I referred to the doc [here](https://docs.frrouting.org/en/latest/bfd.html#configuration)

```
 ! configure a peer with every option possible
 peer 192.168.0.4
  label peer-label
  detect-multiplier 50
  receive-interval 60000
  transmit-interval 3000
  shutdown
``` 

I guess it's yet another case where we should not trust the comments :-) 
Will extend the proposal with all the documented options, thanks!

Reply:
Done